### PR TITLE
Fixes #12063 - Introduce Jetty module for HTTP/2 client dependencies.

### LINKED ISSUE #12063: Introduce Jetty module for HTTP/2 client dependencies
**Jetty version(s)**
12

**Jetty Environment**
core

**Description**
When a web application is proxying and wants to use HTTP/2, it may need to have the HTTP/2 client dependencies provided by the server, rather than the webapp.

There currently is no module for HTTP/2 client, while there is for HTTP/1.

Following the suggestion from #12027, this new Jetty module should download the dependencies on-the-fly.

### PR DESCRIPTION
**Pull request by sbordet**
Made client.mod download dependencies via a [files] section.

Introduced http2-client.mod and http2-client-transport.mod. 
These modules download dependencies via a [files] section. 
They can be used to have the server provide the dependencies in case of a web application proxies request using HTTP/2.

Fixed ContentProvider to set the context ClassLoader before reading the Jetty XML context file, which may reference classes from the web application.

### CODE REVIEW

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java [hunk 1]
```diff
@@ -385,6 +385,10 @@ public ContextHandler createContextHandler(final App app) throws Exception
             // Handle a context XML file
             if (FileID.isXml(path))
             {
+                ClassLoader coreContextClassLoader = Environment.CORE.equals(environment) ? findCoreContextClassLoader(path) : null;
+                if (coreContextClassLoader != null)
+                    Thread.currentThread().setContextClassLoader(coreContextClassLoader);
```
Comment (gregw) [inline jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java:390]: You should remove these lines from 408-409 below

**Review by gregw — CHANGES_REQUESTED** [review]
little niggle

File: jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java (hunk 1 above)
Comment (sbordet) [reply jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java:390]: They are different: they set the classloader on the context, not on the thread.

File: jetty-home/pom.xml [hunk 2]
```diff
@@ -337,7 +341,7 @@
             <configuration>
               <includeGroupIds>org.eclipse.jetty</includeGroupIds>
               <excludeGroupIds>org.eclipse.jetty.orbit,org.eclipse.jetty.http2,org.eclipse.jetty.http3,org.eclipse.jetty.quic,org.eclipse.jetty.websocket,org.eclipse.jetty.ee8.websocket,org.eclipse.jetty.ee9.websocket,org.eclipse.jetty.ee10.websocket,org.eclipse.jetty.fcgi,org.eclipse.jetty.toolchain,org.apache.taglibs</excludeGroupIds>
-              <excludeArtifactIds>jetty-ee8-apache-jsp,jetty-ee9-apache-jsp,jetty-ee10-apache-jsp,jetty-ee8-glassfish-jstl,jetty-ee9-glassfish-jstl,jetty-ee10-glassfish-jstl,jetty-start,jetty-slf4j-impl,javadoc</excludeArtifactIds>
+              <excludeArtifactIds>jetty-alpn-client,jetty-alpn-java-client,jetty-client,jetty-ee8-apache-jsp,jetty-ee9-apache-jsp,jetty-ee10-apache-jsp,jetty-ee8-glassfish-jstl,jetty-ee9-glassfish-jstl,jetty-ee10-glassfish-jstl,jetty-start,jetty-slf4j-impl,javadoc</excludeArtifactIds>
```
Comment (lachlan-roberts) [inline jetty-home/pom.xml]: Are you sure you want to exclude `jetty-client`, it is needed for things like jakarta-websocket-server module, so will save downloading it.

And if they are currently using a 12.0.x release with this then the jar disappears from `jetty-home` in the next point release, then they will have to re-run the `--add-modules` command to download it.

**Review by gregw — DISMISSED** [review]
modulo what @lachlan-roberts asked

File: jetty-home/pom.xml (hunk 2 above)
Comment (sbordet) [reply jetty-home/pom.xml]: Ah yes, the dreaded "the server depends on the client" WebSocket thing 🙄

File: jetty-core/jetty-http2/jetty-http2-client/src/main/config/modules/http2-client.mod [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+[tags]
+client
+http2
+
+[files]
+maven://org.eclipse.jetty/jetty-alpn-client/${jetty.version}/jar|lib/jetty-alpn-client-${jetty.version}.jar
+maven://org.eclipse.jetty/jetty-alpn-java-client/${jetty.version}/jar|lib/jetty-alpn-java-client-${jetty.version}.jar
+maven://org.eclipse.jetty.http2/jetty-http2-client/${jetty.version}/jar|lib/http2/jetty-http2-client-${jetty.version}.jar
+maven://org.eclipse.jetty.http2/jetty-http2-common/${jetty.version}/jar|lib/http2/jetty-http2-common-${jetty.version}.jar
+maven://org.eclipse.jetty.http2/jetty-http2-hpack/${jetty.version}/jar|lib/http2/jetty-http2-hpack-${jetty.version}.jar
```
Comment (lachlan-roberts) [inline jetty-core/jetty-http2/jetty-http2-client/src/main/config/modules/http2-client.mod]: The `http2-hpack` and `http2-common` jars are already in jetty-home, so these are not needed.

Comment (sbordet) [reply jetty-core/jetty-http2/jetty-http2-client/src/main/config/modules/http2-client.mod]: Done.

**Pull request merged**